libultraship/libultraship/SohHooks.cpp

```diff
@@ -1,14 +1,25 @@
 #include "SohHooks.h"
-#include <map>
-#include <string>
 #include <vector>
-#include <stdarg.h>
+#include <cstdarg>
 #include <iostream>
 
-std::map<std::string, std::vector<HookFunc>> listeners;
-std::string hookName;
-std::map<std::string, void*> initArgs;
-std::map<std::string, void*> hookArgs;
+std::unordered_map<HookTable, std::vector<void (*)(HookEvent)>> listeners;
+std::unordered_map<unsigned long, void*> hookArgs;
+HookTable hookEvent = NULL_HOOK;
+
+unsigned long hash(const char* str) {
+    unsigned long hash = 5381;
+    int c;
+
+    while (c = *str++)
+        hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+
+    return hash;
+}
```

Comment:
You should try to see if `std::hash<std::string_view>()(...)` is faster. If you can change `const char *` to `std::string_view` the whole way, it also does not need to find the end of the string at runtime.